File: scripts/seed-bbmp-street-index.js
#!/usr/bin/env node
// scripts/seed-bbmp-street-index.js
// Bulk-insert the 9,913-row Bengaluru street index from
// tmp/bbmp-street-index.json into regulatory_data.bbmp_street_index.
// Reads DATABASE_URL from backend/.env so the same pg pool the live API uses
// also gets the seed — no second connection string to manage.
//
// Usage:
//   node scripts/seed-bbmp-street-index.js
//
// Idempotent: the unique index (lower(street_name_en), ward_no, page_number)
// silently drops duplicates via ON CONFLICT DO NOTHING.

'use strict';

const fs = require('node:fs');
const path = require('node:path');
const { Pool } = require('pg');

require('dotenv').config({ path: path.join(__dirname, '..', 'backend', '.env') });

const JSON_PATH = path.join(__dirname, '..', 'tmp', 'bbmp-street-index.json');
const BATCH_SIZE = 500;

async function main() {
  if (!process.env.DATABASE_URL) {
    console.error('ERROR: DATABASE_URL missing from backend/.env');
    process.exit(1);
  }
  if (!fs.existsSync(JSON_PATH)) {
    console.error(`ERROR: missing ${JSON_PATH} — run build-bbmp-street-index.py first`);
    process.exit(1);
  }

  const rows = JSON.parse(fs.readFileSync(JSON_PATH, 'utf-8'));
  console.log(`Loaded ${rows.length} street entries from ${JSON_PATH}`);

  const pool = new Pool({
    connectionString: process.env.DATABASE_URL,
    ssl: { rejectUnauthorized: false },
    max: 4,
  });

  const before = await pool.query('SELECT COUNT(*)::int AS n FROM regulatory_data.bbmp_street_index');
  console.log(`Before seed: ${before.rows[0].n} rows already present`);

  let inserted = 0;
  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const chunk = rows.slice(i, i + BATCH_SIZE);
    const valuesSql = [];
    const params = [];
    for (const r of chunk) {
      const aro = (r.aro_section || '').replace(/ /g, ' ').trim() || null;
      const idx = params.length;
      params.push(r.street_name_en, r.ward_no ?? null, r.page_number, aro, r.row_excerpt ?? null);
      valuesSql.push(`($${idx + 1}, $${idx + 2}, $${idx + 3}, $${idx + 4}, $${idx + 5})`);
    }
    const sql = `
      INSERT INTO regulatory_data.bbmp_street_index
        (street_name_en, ward_no, page_number, aro_section, row_excerpt)
      VALUES ${valuesSql.join(', ')}
      ON CONFLICT (lower(street_name_en), ward_no, page_number) DO NOTHING
      RETURNING 1`;
    const result = await pool.query(sql, params);
    inserted += result.rowCount;
    process.stdout.write(`\r  Batch ${Math.floor(i / BATCH_SIZE) + 1}/${Math.ceil(rows.length / BATCH_SIZE)}  (+${result.rowCount} new, ${inserted} total inserted)`);
  }
  process.stdout.write('\n');

  const after = await pool.query('SELECT COUNT(*)::int AS n FROM regulatory_data.bbmp_street_index');
  console.log(`After seed: ${after.rows[0].n} rows`);
  await pool.end();
}

main().catch((err) => {
  console.error('Seed failed:', err);
  process.exit(1);
});

File: scripts/seed-bbmp-street-index.py
#!/usr/bin/env python3
"""Convert tmp/bbmp-street-index.json into a sequence of batched INSERT SQL
files the agent can apply via the Supabase MCP execute_sql tool.

Why batched files vs one big migration:
    - 9,913 rows in one INSERT exceeds Supabase SQL editor and MCP payload
      limits in practice.
    - 500 rows per batch ≈ ~80 KB SQL per file, well under any limit.
    - File names sort lexicographically so the agent can apply them in order.

Output: tmp/bbmp-street-seed/batch-NNN.sql (20 files for 9,913 rows).
"""
from __future__ import annotations
import json
import re
import sys
from pathlib import Path

TMP = Path(__file__).resolve().parent.parent / "tmp"
IN_FILE = TMP / "bbmp-street-index.json"
OUT_DIR = TMP / "bbmp-street-seed"
BATCH_SIZE = 500


def sql_quote(value) -> str:
    """Single-quote-safe SQL literal for text values."""
    if value is None:
        return "NULL"
    if isinstance(value, (int, float)):
        return str(value)
    escaped = str(value).replace("'", "''")
    return f"'{escaped}'"


def main() -> int:
    if not IN_FILE.exists():
        print(f"ERROR: missing {IN_FILE} — run build-bbmp-street-index.py first", file=sys.stderr)
        return 1

    data = json.loads(IN_FILE.read_text(encoding="utf-8"))
    OUT_DIR.mkdir(parents=True, exist_ok=True)

    # Wipe stale batches so re-runs are clean.
    for old in OUT_DIR.glob("batch-*.sql"):
        old.unlink()

    total_batches = (len(data) + BATCH_SIZE - 1) // BATCH_SIZE

    for batch_idx in range(total_batches):
        start = batch_idx * BATCH_SIZE
        end = min(start + BATCH_SIZE, len(data))
        chunk = data[start:end]

        rows_sql = []
        for r in chunk:
            # Normalise ARO section (drop \xa0 non-breaking spaces).
            aro = (r.get("aro_section") or "").replace("\xa0", " ").strip() or None
            cols = [
                sql_quote(r["street_name_en"]),
                sql_quote(r.get("ward_no")),
                sql_quote(r["page_number"]),
                sql_quote(aro),
                sql_quote(r.get("row_excerpt")),
            ]
            rows_sql.append(f"  ({', '.join(cols)})")

        sql = (
            "INSERT INTO regulatory_data.bbmp_street_index\n"
            "  (street_name_en, ward_no, page_number, aro_section, row_excerpt)\n"
            "VALUES\n"
            + ",\n".join(rows_sql)
            + "\nON CONFLICT (lower(street_name_en), ward_no, page_number) DO NOTHING;\n"
        )

        out_path = OUT_DIR / f"batch-{batch_idx + 1:03d}.sql"
        out_path.write_text(sql, encoding="utf-8")

    print(f"Wrote {total_batches} batches ({len(data)} rows) to {OUT_DIR}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
